_drafts/printf-on-embedded.md

```diff
@@ -0,0 +1,566 @@
+can be found here:
+
+<https://mcuoneclipse.com/2016/10/17/tutorial-using-single-wire-output-swo-with-arm-cortex-m-and-eclipse/>
+
+The debug adapter and software does need to support the protocol. Most debug
+adapters do support it, however (ST-Link's, JLINK, DAPLink, etc).
+
+Pros:
+
+- generally pretty fast, similar performance to UART (speed is both target and
+  probe hardware dependent)
+- simple implementation
+
+Cons:
+
+- output only
+- uses a pin on the target
+- not all chips or probes support it
+
+### Real Time Transfer (RTT)
+
+<https://wiki.segger.com/RTT>
+
+Segger's RTT reads/writes into a buffer in RAM while the chip is running. It can
+be pretty fast compared to even a fast UART, and is _much_ more efficient than
+Semihosting without requiring any extra pins or hardware.
+
+It also supports both output (target -> host) and input (host -> target).
+
+Segger provides the RTT target implementation, as well as a few tools for using
+RTT on the host. OpenOCD and PyOCD also support the protocol.
+
+You can download the implementation here:
+
+<https://www.segger.com/products/debug-probes/j-link/technology/about-real-time-transfer/#downloading-rtt-target-package>
+
+Zephyr also maintains a copy here, if you're curious about perusing it (or want
+to use it as a git submodule):
+
+<https://github.com/zephyrproject-rtos/segger/tree/master/SEGGER>
+
+The downside is it requires a debug adapter to operate, since it relies on
+reading and writing to memory.
+
+It's also important to understand the implications of enabling this in
+production; be sure your device doesn't get into a bad state if there's no debug
+adapter connected (i.e. when no debugger is attached, drop data instead of
+backing up the RTT buffers).
+
+I'm personally a big fan of RTT over semihosting- it's usually going to be a
+strict improvement if you don't need file I/O.
+
+Pros:
+
+- fast!
+- simple on-device implementation
+- supported out of the box on Zephyr RTOS, among other platforms
+
+Cons:
+
+- requires a debugger with support for RTT
+
+## `flush()`
+
+That's all for now! Hopefully this article provided a little window into the
+various ways `printf` can be used on small microcontrollers (or other unhosted
+environments). It's a surprisingly subtle corner of embedded development, and
+there's a few pitfalls that should be watched for, especially around concurrency
+or performance.
+
+The links below should provide a lot of additional information if you're curious
+about exploring the topic further.
+
+And as always, if you have any questions or comments, feel free to reach out-
+either in the comments here or in the Interrupt community Slack!
+
+<!-- Interrupt Keep START -->
+
+{% include newsletter.html %}
+
+{% include submit-pr.html %}
+
+<!-- Interrupt Keep END -->
+
+{:.no_toc}
+
+## References
+
+<!-- prettier-ignore-start -->
+
+- [Interrupt: Introduction to ARM Semihosting]({% post_url 2021-02-16-arm-semihosting %})
```

Comment:
Can we add these links directly to the spots they're referenced? Might mike it easier for people to dive deeper as they read.

Reply:
they're all just copied from above, all the links are inline :grimacing: personally I prefer having bare urls, do we prefer footnote links or hyperlinks?

Reply:
I don't think we have a hard preference. I like the superscript links personally, but I think it's fine as is.